neps/nep-0366.md

```diff
@@ -0,0 +1,203 @@
+
+A `message` is formed in the following format:
+```rust
+struct DelegateActionMessage {
+    signed_delegate_action: SignedDelegateAction
+}
+```
+
+The next set of security concerns are addressed by this format:
+ - `sender_id` is included to ensure that the relayer set correct `transaction.receiver_id`.
+ - `max_block_height` is included to ensure that the `DelegateAction` isn't expired.
+ - `nonce` is included to ensure that the `DelegateAction` can't be replayed again.
+ - `public_key` and `sender_id` are needed to ensure that on the right account, work across rotating keys and fetch correct `nonce`.
+
+The permissions are verified based on the variant of `public_key`:
+ - `AccessKeyPermission::FullAccess`, all actions are allowed.
+ - `AccessKeyPermission::FunctionCall`, only a single `FunctionCall` action is allowed in `actions`.
+    - `DelegateAction.receiver_id` must match to the `account[public_key].receiver_id`
+    - `DelegateAction.actions[0].method_name` must be in the `account[public_key].method_names`
+
+***Outcomes***:
+- If the `signature` matches the receiver's account's `public_key`, a new receipt is created from this account with a set of `ActionReceipt { receiver_id, action }` for each action in `actions`.
+
+
+***Recommendations***
+- Because the User doesn't trust the Relayer, the User should verify whether the Relayer has submitted the `DelegateAction` and the execution result.
+
+#### Errors
+
+- If the Sender's account doesn't exist
+```rust
+/// Happens when TX receiver_id doesn't exist
+AccountDoesNotExist
+```
+
+- If the `signature` does not match the data and the `public_key` of the given key, then the following error will be returned
+```rust
+/// Signature does not match the provided actions and given signer public key.
+DelegateActionInvalidSignature
+```
+
+- If the `sender_id` doesn't match the `tx.receiver_id`
+```rust
+/// Receiver of the transaction doesn't match Sender of the delegate action
+DelegateActionSenderDoesNotMatchReceiver
+```
+
+- If the current block is equal or greater than `max_block_height`
+```rust
+/// Delegate action has expired
+DelegateActionExpired
+```
+
+- If the `public_key` does not exist for Sender account
+```rust
+/// The given public key doesn't exist for Sender account
+DelegateActionAccessKeyError
+```
+
+- If the `nonce` does match the `public_key` for the `sender_id`
+```rust
+/// Nonce must be greater sender[public_key].nonce
+DelegateActionInvalidNonce
+```
+
+- If `nonce` is too large
+```rust
+/// DelegateAction nonce is larger than the upper bound given by the block height (block_height * 1e6)
+DelegateActionNonceTooLarge
+```
+
+- If the list of delegated actions contains another `DelegateAction`
+```rust
+/// DelegateAction actions contain another DelegateAction
+DelegateActionCantContainNestedOne
+```
+
+- If the list of Transaction actions contains several `DelegateAction`
+```rust
+/// There should be the only one DelegateAction
+DelegateActionMustBeOnlyOne
```

Comment:
Curious why this is the case? I don't see the reasoning in this doc; sorry if I skimmed over it. This seems like a benefit to have to be able to schedule multiple meta transactions within a single transaction, to minimize costs for the relayer.

If you actually want this to be the only thing executed within the transaction, should regular actions not be allowed with a delegate action? Currently, the draft implementation seems to allow other actions with a delegate action, and I just want to make sure this is intended

Reply:
I added this case for two reasons:
1. At this moment all actions in the  transaction are executed in one by one. But in case of `DelegateAction`, they will be executed in parallel (inner actions will be executed in parallel). I think such behavior might be confusing for developer/users.
2. For code simplification. If we allow several `DelegateAction` in a transaction, we should decide how limits should be handled. For example there is the limit for total number of actions in the transactions. Should it be shared between `DelegateAction` and the transaction or we need to define a new limit for number of actions in `DelegateAction`?